README.md
go-quantize converts an arbitrary sequence of floating point inputs into a well defined output space.

# Use Case

One use case for a numeric quantization function is to normalize inputs into a more reasonable output space. If a given function or set of measurements produces values too small or densely packed for human "consumption"/analysis, quantizing them into an integer space would allow a human to "eyeball" the distinctions between these values more sensible.

# Caveats
Quantization is a *lossy* conversion. The quantization function is in effect a hash function and depending on your inputs and output parameters, multiple input values may be mapped to the same output value. Be aware of this and do not rely on perfect fidelity of input->output mappings.

# Example

For the purposes of this example, assume that `GenFloats` is a function that produces values in the interval [-1,1].


```go
import "fmt"

d := Domain{
  Min:  -5.0,
  Max:  5.0,
  Step: 1.0,
}

fs := GenFloats(5, 5, 8675309)

v := d.Quantize(fs)

fmt.Printf("Output Domain: %+v\n", d)
fmt.Printf("Input: %v\n", fs)
fmt.Printf("Output: %v\n", v)
```

The return value of `Quantize()` is an array of float64 values, normalized to the given Domain: in the case of the example, there will be no values smaller than -5, no values greater than 5.0, and all values will be integer increments between the two end points: -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, and 5.


CHANGELOG.md
# Changelog

## [Unreleased]
### Changed

[Unreleased]: https://github.com/therealfakemoot/go-quantize/compare/v0.0.1...HEAD


test_q.py
import unittest

from q import Q

class TestSimpleDomain(unittest.TestCase):
    def testFullRange(self):
        steps = range(-5, 6, 1)
        steps = list(steps)

        fs = [-1, -.8, -.6, -.4, -.2, 0, .2, .4, .6, .8, 1]

        expected = [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5]
        actual = Q(steps, fs)
        self.assertEqual(expected, list(actual))

class TestBalancedDomains(unittest.TestCase):
    def testOddStepCount(self):
        steps = range(-10, 21, 3)
        steps = list(steps)

        fs = [-1, -.8, -.6, -.4, -.2, 0, .2, .4, .6, .8, 1]

        expected = [-10, -7, -4, -1, 2, 5, 8, 11, 14, 17, 20]
        actual = Q(steps, fs)
        self.assertEqual(expected, list(actual))

    def testEvenStepCount(self):
        steps = range(-10, 21, 3)
        steps = list(steps)

        fs = [-1, -.8, -.6, -.4, -.2, 0, .2, .4, .6, .8, 1]

        expected = [-10, -7, -4, -1, 2, 5, 8, 11, 14, 17, 20]
        actual = Q(steps, fs)
        self.assertEqual(expected, list(actual))


class TestUnbalancedDomains(unittest.TestCase):
    def testPositiveHeavyDomain(self):
        steps = list(range(-10, 21, 4))

        fs = [-1, -.75, -.5, -.25, .25, .5, .75, 1]

        expected = [-10, -6, -2, 2, 6, 10, 14, 18]
        actual = Q(steps, fs)
        self.assertEqual(expected, list(actual))

    def testNegativeHeavyDomain(self):
        steps = list(range(-20, 11, 4))

        fs = [-1, -.75, -.5, -.25, .25, .5, .75, 1]

        expected = [-20, -16, -12, -8, -4, 0, 4, 8]
        actual = Q(steps, fs)
        self.assertEqual(expected, list(actual))

if __name__ == '__main__':
    unittest.main()


domain_test.go
package quantize

import (
	"reflect"
	"testing"
)

var (
	seed    int64 = 8675309
	edge          = -20.0
	mapMin        = 5.0
	mapMax        = 20.0
	mapStep       = 5.0
)

func buildTestDomains(edge float64) []Domain {
	var ret []Domain

	for x := edge; x < 0.0; x++ {
		ret = append(ret, Domain{Min: x, Max: -x})
	}
	return ret
}

func TestKnownValues(t *testing.T) {
	input := Domain{Min: -1, Max: 1}
	t.Run("Balanced Domains", func(t *testing.T) {
		t.Run("[-5,5]", func(t *testing.T) {
			d := Domain{Min: -5, Max: 5}
			fs := []float64{-1.0, -0.8, -0.6, -0.4, -0.2, 0, 0.2, 0.4, 0.6, 0.8, 1}
			quantized := QuantizeAll(fs, input, d)
			expected := []float64{-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5}

			if !reflect.DeepEqual(quantized, expected) {
				t.Errorf("Expected: %v, got %v", expected, quantized)
			}
		})
		t.Run("[-100,100]", func(t *testing.T) {
			d := Domain{Min: -100, Max: 100}
			fs := []float64{-1.0, -0.8, -0.6, -0.4, -0.2, 0, 0.2, 0.4, 0.6, 0.8, 1}
			quantized := QuantizeAll(fs, input, d)
			expected := []float64{-100, -80, -60, -40, -20, 0, 20, 40, 60, 80, 100}

			if !reflect.DeepEqual(quantized, expected) {
				t.Errorf("Expected: %v, got %v", expected, quantized)
			}
		})
		t.Run("[-2500,2500]", func(t *testing.T) {
			t.Skip()
			d := Domain{Min: -2500, Max: 2500}
			fs := []float64{-1.0, -0.8, -0.6, -0.4, -0.2, 0, 0.2, 0.4, 0.6, 0.8, 1}
			quantized := QuantizeAll(fs, input, d)
			expected := []float64{-100, -80, -60, -40, -20, 0, 20, 40, 60, 80, 100}

			if !reflect.DeepEqual(quantized, expected) {
				t.Errorf("Expected: %v, got %v", expected, quantized)
			}
		})

	})

	d := Domain{Min: -5, Max: 5}
	fs := []float64{-1.0, -0.8, -0.6, -0.4, -0.2, 0, 0.2, 0.4, 0.6, 0.8, 1}
	quantized := QuantizeAll(fs, input, d)
	expected := []float64{-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5}

	if !reflect.DeepEqual(quantized, expected) {
		t.Errorf("Expected: %v, got %v", expected, quantized)
	}
}


quantize_test.go
package main

import (
	"testing"
)

func MinFloat(ints []float64) float64 {
	var min float64

	for _, v := range ints {
		if v < min {
			min = v
		}
	}

	return min
}

func MaxFloat(ints []float64) float64 {
	var max float64

	for _, v := range ints {
		if v > max {
			max = v
		}
	}

	return max
}

func prepTests(edge float64) []Domain {
	var ret []Domain

	for x := edge; x < 0.0; x++ {
		ret = append(ret, Domain{Min: x, Max: -x, Step: 1})
	}
	return ret
}

func TestDomainMin(t *testing.T) {
	domains := prepTests(-20)

	for _, d := range domains {

		for i := 5.0; i <= 20.0; i += 5 {
			fs := GenFloats(i, i, 8675309)

			quantized := d.Quantize(fs)

			t.Logf("Domain: %+v", d)
			t.Logf("GenFloats(%f, %f)", i, i)
			t.Logf("Quantized Values: %v", quantized)

			min := MinFloat(quantized)

			if min < d.Min {
				t.Errorf("Domain Minimum (%f) exceeded: %f\n", d.Min, min)
			}
		}
	}

}

func TestDomainMax(t *testing.T) {
	domains := prepTests(-20)

	for _, d := range domains {

		for i := 5.0; i <= 20.0; i += 5 {
			fs := GenFloats(i, i, 8675309)

			quantized := d.Quantize(fs)

			t.Logf("Domain: %+v", d)
			t.Logf("GenFloats(%f, %f)", i, i)
			t.Logf("Quantized Values: %v", quantized)
			max := MaxFloat(quantized)

			if max > d.Max {
				t.Errorf("Domain Maximum (%f) exceeded: %f", d.Max, max)
			}
		}
	}
}

func TestDomainSteps(t *testing.T) {
	d := Domain{
		Min:  -5,
		Max:  5,
		Step: 1,
	}

	steps := d.Steps()

	if len(steps) != 11 {
		t.Errorf("Expected 11 steps, got: %d", len(steps))
	}
}


helpers/lineup.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker

points = lambda x,y,z: [i/10 for i in range(x*10,(y*10)+1,int(z*10))]

def setupBase(ax):
    ax.spines['right'].set_color('none')
    ax.spines['left'].set_color('none')
    ax.yaxis.set_major_locator(ticker.NullLocator())
    ax.spines['top'].set_color('none')
    ax.xaxis.set_ticks_position('bottom')
    ax.tick_params(which='major', width=1.00)
    ax.tick_params(which='major', length=5)
    ax.tick_params(which='minor', width=0.75)
    ax.tick_params(which='minor', length=2.5)
    ax.set_ylim(0, 1)
    ax.patch.set_alpha(0.0)

def setupDomain(ax):
    ax.spines['right'].set_color('none')
    ax.spines['left'].set_color('none')
    ax.yaxis.set_major_locator(ticker.NullLocator())
    ax.spines['bottom'].set_color('none')
    ax.xaxis.set_ticks_position('bottom')
    ax.tick_params(which='major', width=1.00)
    ax.tick_params(which='major', length=5)
    ax.tick_params(which='minor', width=0.75)
    ax.tick_params(which='minor', length=2.5)
    ax.set_ylim(0, 1)
    ax.patch.set_alpha(0.0)

plt.figure(figsize=(8,6))

p = points(-1, 1, 0.1)
y = [0 for i in p]

base = plt.subplot(8, 1, 1)
setupBase(base)
base.set_xlim(min(p), max(p))

# domain = plt.subplot(9, 1, 1)
# setupDomain(domain)
# domain.set_xlim(-5, 5)

fs = [1.9052595476929043e-65, -0.24547524871149487, -0.14958647968356706, 0.2524271844660192, 0.07543249830197025, -0.2358464181549405, -2.0927125031917226e-65, -0.31351632106756133, 0.43597426984697757, 0.06148867313915843]

plt.plot(fs, y[:len(fs)], "r|")
plt.savefig('/var/www/idle.ndumas.com/line.svg')


helpers/q.py
import math

def abs_diff(x, y):
    return abs(min(x, y) - max(x,y))

midpoint = lambda x: (x[0] + x[-1])/2.0

def balanceSteps(steps):
    m = midpoint(steps)
    shift = abs_diff(m, 0)

    if abs(min(steps)) < abs(max(steps)):
        shifted = [step - shift for step in steps]
    else:
        shifted = [step + shift for step in steps]

    return shifted

def Q(steps, fs):
    low, high = min(steps), max(steps)
    for f in fs:
        yield (f-(-1))*(high-low)/(1-(-1)) + low


if __name__ == '__main__':

    # The output domain is the set of possible values that inputs can be mapped to.
    output_domain = list(range(-30, 16, 1))

    # This is an example input set, specifically crafted to match the output domain above.
    # -1 should be mapped to -5, -0.8 should map to -4, etc.
    fs = [-1, -.8, -.6, -.4, -.2, 0, .2, .4, .6, .8, 1]

    # fs = [1.9052595476929043e-65,-0.24547524871149487,-0.14958647968356706,0.2524271844660192,0.07543249830197025,-0.2358464181549405,-2.0927125031917226e-65,-0.31351632106756133,0.43597426984697757,0.06148867313915843]

    o = Q(output_domain, fs)
    # print(o)
